src/QMCDrivers/WFOpt/EngineHandle.h

```diff
@@ -92,20 +93,24 @@ class DescentEngineHandle : public EngineHandle
   void takeSample(const std::vector<FullPrecReal>& energy_list,
                   const RecordArray<Value>& dlogpsi_array,
                   const RecordArray<Value>& dhpsioverpsi_array,
-                  int ib) override
+                  int base_sample_index) override
   {
-    der_rat_samp[0] = 1.0;
-    le_der_samp[0]  = energy_list[ib];
-
-    int num_params = der_rat_samp.size() - 1;
-    for (int j = 0; j < num_params; j++)
+    int current_batch_size = dlogpsi_array.nentry();
+    for (int local_index = 0; local_index < current_batch_size; local_index++)
     {
-      der_rat_samp[j + 1] = static_cast<FullPrecValue>(dlogpsi_array.getValue(j, ib));
-      le_der_samp[j + 1]  = static_cast<FullPrecValue>(dhpsioverpsi_array.getValue(j, ib)) +
-          le_der_samp[0] * static_cast<FullPrecValue>(dlogpsi_array.getValue(j, ib));
+      der_rat_samp[0] = 1.0;
+      le_der_samp[0]  = energy_list[local_index];
+
+      int num_params = der_rat_samp.size() - 1;
+      for (int j = 0; j < num_params; j++)
+      {
+        der_rat_samp[j + 1] = static_cast<FullPrecValue>(dlogpsi_array.getValue(j, local_index));
+        le_der_samp[j + 1]  = static_cast<FullPrecValue>(dhpsioverpsi_array.getValue(j, local_index)) +
+            le_der_samp[0] * static_cast<FullPrecValue>(dlogpsi_array.getValue(j, local_index));
+      }
+      int ip = omp_get_thread_num();
+      engine_.takeSample(ip, der_rat_samp, le_der_samp, le_der_samp, 1.0, 1.0);
```

Comment:
How this relates to all the replica based accesses in the DescentEngine itself is really cryptic. Since this oenpmp thread id which isn't guaranteed to be any particular number especially in the batched context yet it is used as the replica id in the engine, which seems like it is very important.